cmd/erasure-server-pool.go

```diff
@@ -1415,69 +1415,65 @@ func (z *erasureServerPools) Walk(ctx context.Context, bucket, prefix string, re
 type HealObjectFn func(bucket, object, versionID string) error
 
 func (z *erasureServerPools) HealObjects(ctx context.Context, bucket, prefix string, opts madmin.HealOpts, healObject HealObjectFn) error {
-	// If listing did not return any entries upon first attempt, we
-	// return `ObjectNotFound`, to indicate the caller for any
-	// actions they may want to take as if `prefix` is missing.
-	err := toObjectErr(errFileNotFound, bucket, prefix)
-	for _, erasureSet := range z.serverPools {
-		for _, set := range erasureSet.sets {
-			var entryChs []FileInfoVersionsCh
-			var mu sync.Mutex
-			var wg sync.WaitGroup
-			for _, disk := range set.getOnlineDisks() {
-				disk := disk
-				wg.Add(1)
-				go func() {
-					defer wg.Done()
-					entryCh, err := disk.WalkVersions(ctx, bucket, prefix, "", true, ctx.Done())
-					if err != nil {
-						// Disk walk returned error, ignore it.
-						return
-					}
-					mu.Lock()
-					entryChs = append(entryChs, FileInfoVersionsCh{
-						Ch: entryCh,
-					})
-					mu.Unlock()
-				}()
-			}
-			wg.Wait()
-
-			entriesValid := make([]bool, len(entryChs))
-			entries := make([]FileInfoVersions, len(entryChs))
+	errCh := make(chan error)
+	go func() {
+		defer close(errCh)
 
-			for {
-				entry, quorumCount, ok := lexicallySortedEntryVersions(entryChs, entries, entriesValid)
-				if !ok {
-					break
+		for _, erasureSet := range z.serverPools {
+			for _, set := range erasureSet.sets {
+				disks, _ := set.getOnlineDisksWithHealing()
+				if len(disks) == 0 {
+					errCh <- errors.New("HealObjects: No non-healing disks found")
+					return
 				}
 
-				// Remove empty directories if found - they have no meaning.
-				// Can be left over from highly concurrent put/remove.
-				if quorumCount > set.setDriveCount/2 && entry.IsEmptyDir {
-					if !opts.DryRun && opts.Remove {
-						set.deleteEmptyDir(ctx, bucket, entry.Name)
+				deepScan := opts.ScanMode == madmin.HealDeepScan
+				healEntry := func(entry metaCacheEntry) {
+					if entry.isDir() {
+						return
+					}
+					fivs, err := entry.fileInfoVersions(bucket)
+					if err != nil {
+						errCh <- err
+						return
+					}
+					waitForLowHTTPReq(globalHealConfig.IOCount, globalHealConfig.Sleep)
```

Comment:
Since we are often dealing with remote disks a dynamic wait may be more appropriate?

Reply:
A dynamic wait is already inside healObject() this is just to honor what customer might have set.